sqlalchemy-stubs/sql/schema.pyi

```diff
@@ -545,7 +545,11 @@ class Index(DialectKWArgs, ColumnCollectionMixin, SchemaItem):
     unique: bool = ...
     expressions: List[Column[Any]] = ...
     def __init__(
-        self, name: str, *expressions: ColumnClause[Any], **kw: Any
+        self,
+        name: str,
+        # TODO This is technically DDLConstraintColumnRole, do we use roles for types?
+        *expressions: Union[ColumnClause[Any], ColumnElement[Any], TextClause],
```

Comment:
From my conversation with @CaselIT and @zzzeek, we can't use roles for types. However, I don't remember why off the top of my head.

Reply:
I don't actually remember.

Reply:
it's not typically very applicable to schema things like Index, but to support ORM use cases, there's the single case of an ORM mapped class that is user defined and therefore does not suit any kind of role directly, it has to go through inspect() first:



    class MyClass(Base):
         # ...

    stmt = select(MyClass)    # <---- no role

that's pretty much the only real world use case right now.

Reply:
OK so adding more completeness, anything that either responds to inspect(), or anything that has a `__clause_element__()` method, which we call to get the thing.  However, real-world use of `__clause_element__()` is always on an object that suits a role in any case.

Reply:
I would say for now that anything that accepts column elements / text like Index above and does not accept FROM objects is probably fine to use roles.    Really the thing with ORM classes being sent is `select()`, `select().select_from()`, and ...maybe really not much more than that actually.

Reply:
`join` also comes to mind

Reply:
there is a separate ORMJoin construct, though I guess sqlalchemy.sql.join(ClassA, ClassB) will now do something (I dont think it did in 1.3 but I haven't tried).

Reply:
also `select(ClassA).join(ClassB)`. I don't remember also, I used it also on 1.3, but I don't remember if it was with `ClassFoo` or `ClassFoo.__table__`

Reply:
basically things that accept FromClause types of roles will also need to accept ORM mapped classes.  otheriwse I think roles usually work.

Reply:
I've adapted this to the correct role.